### PR TITLE
Add new transitions report

### LINKED ISSUE #12865: New report "Transitions" available under the "Actions" category
Transitions is a fantastic report and very insightful: https://matomo.org/docs/transitions/

The idea of this issue is to make Transitions a first class citizen of Matomo and visible to all users.

### Suggested steps

* Add a new report "Transitions" below "Contents"
* The transition report is then loaded in the page (not in popover). It shows the transition for the #1 most visited page URL on the site.
* A "related report" is available to switch to "Transitions (Page titles)" which shows the #1 most visited page title  

### Out of scope but related

* in the footer, ideally we show the "Export data" icon https://github.com/matomo-org/matomo/issues/6032
* in the footer, ideally we show "Export as image" icon https://github.com/matomo-org/matomo/issues/6248

### PR DESCRIPTION
**Pull request by tsteur**
As you can maybe see in the first commit I tried to first make it work as a report  with a visualization and including related report etc but couldn't get it to work. There were various challenges. So made it a widget instead. Was still quite tricky to get it to work and is a bit messy since it was all hard adjusted to be a popover.

Looks like this:
image

No data screen:
image

Menu: 
image

There is no search in the label field for now. It's only our regular select. Should do for now I'd say and at some point need to add support for autocomplete or so.

I'll try to add some UI tests if that's looking good @mattab 

fix #12865

### CODE REVIEW

File: plugins/Transitions/API.php [hunk 1]
```diff
@@ -617,4 +617,22 @@ protected function processTransitionsToFollowingPages($type, $actions)
...
+     * @return DataTable
+     */
+    public function getPageURLTransitions($idSite, $period, $date, $segment = false)
+    {
+        $table = new DataTable();
+
+        $table->addRowFromArray(array(Row::COLUMNS => array('nb_visits' => 5)));
+
+        return $table;
+    }
```
Comment (diosmosis) [inline plugins/Transitions/API.php]: Should this be here?

Comment (tsteur) [reply plugins/Transitions/API.php]: will remove it, was auto generated 👍

**Comment by diosmosis** [comment]
Testing locally noticed these issues:

1) the report doesn't seem to be centered, might look better if it was

image

2) If I click in the transitions report, I see a popup like this

image

  and then the report is empty

image

3) the select dropdown is under the report

image

**Comment by diosmosis** [comment]
@mattab UI needs approval

File: plugins/Transitions/angularjs/transitionswitcher/transitionswitcher.controller.js [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+                        }
+                    }
+                    self.onActionNameChange(self.actionName);
+                }
+
+                if (!self.actionName || self.actionNameOptions.length === 0) {
+                    self.isEnabled = false;
+                    self.actionName = '';
+                    self.actionNameOptions.push({key: '', value: translate('CoreHome_ThereIsNoDataForThisReport')});
+                }
```
Comment (diosmosis) [inline plugins/Transitions/angularjs/transitionswitcher/transitionswitcher.controller.js]: Not super important, but I think this method could be split up for better readability, if desired.

Comment (tsteur) [reply plugins/Transitions/angularjs/transitionswitcher/transitionswitcher.controller.js]: will do 👍

File: plugins/Transitions/lang/en.json [hunk 3]
```diff
@@ -2,11 +2,14 @@
...
         "BouncesInline": "%s bounces",
         "DirectEntries": "Direct Entries",
+        "Transitions": "Transitions",
         "ErrorBack": "Go back to the previous action",
         "ExitsInline": "%s exits",
         "NumPageviews": "%s pageviews",
         "NumDownloads": "%s downloads",
         "NumOutlinks": "%s outlinks",
+        "TopX": "Top %s labels",
+        "AvailableInOtherReports": "Did you know? Transitions are also available as a row action in the following reports:",
```
[the remark this answers is not in the document]
Comment (tsteur) [reply plugins/Transitions/lang/en.json]: will add it: 👍

**Comment by tsteur** [comment]
> the report doesn't seem to be centered, might look better if it was

I didn't manage to center it but I actually think it's fine like this anyway. Feel free to give it a try. It's so hard to do much there I think cause the transitions JS is quite complex and built for popup etc.

> If I click in the transitions report, I see a popup like this

had actually no idea it was clickable... i think the JS is not able to render 2 transition reports on the same page or it's trying to alter something. 
It should more or less work now. If there's still a problem I will need to disable the onclick functionality in this report.

> the select dropdown is under the report

should be fixed

also applied the other feedback @diosmosis

**Comment by diosmosis** [comment]
LGTM just need UX approval from @mattab & UI tests

**Comment by tsteur** [comment]
Tests should be fixed now.

**Comment by diosmosis** [comment]
Was going to merge, but the transitions widget looks a bit off:

image

I think something in the code is also causing the UI test on travis to have issues (works fine for me locally): https://builds-artifacts.matomo.org/matomo-org/matomo/12865/31592/UIIntegrationTest_dashboard3.png

Think it might have something to do w/ the width of an element.

**Comment by tsteur** [comment]
@diosmosis I think I fixed the remaining issues this morning

**Comment by mattab** [comment]
Looks really good overall, it's nice to know more users will discover this feature.
Feedback:

* In the footer message eg. ` Simply hover a row in any of these reports and click on the transition icon to launch it. ` could maybe add the Transitions icon in the text itself.
* Add a "Help text" for the report, that 1) gives a quick explanation of what the report is, and 2) links to: https://matomo.org/docs/transitions/ 
* When viewing the transitions for a site, and changing site, then I get error:

```
Please specify a value for 'actionName'.#0 /core/API/Proxy.php(157): Piwik\API\Proxy->getRequestParametersArray(Array, Array)#1 /core/Context.php(28): Piwik\API\Proxy->Piwik\API\{closure}()#2 /core/API/Proxy.php(323): Piwik\Context::executeWithQueryParameters(Array, Object(Closure))#3 /core/API/Request.php(263): Piwik\API\Proxy->call('\\Piwik\\Plugins\\...', 'getTransitionsF...', Array)#4 /plugins/API/Controller.php(41): Piwik\API\Request->process()#5 [internal function]: Piwik\Plugins\API\Controller->index()#6 /core/FrontController.php(564): call_user_func_array(Array, Array)#7 /core/FrontController.php(152): Piwik\FrontController->doDispatch('API', false, Array)#8 /core/dispatch.php(34): Piwik\FrontController->dispatch()#9 /index.php(27): require_once('/home/matt/dev/...')#10 {main}
``` 

* in case it's possible, would be nice if we could re-draw the Transitions report when the browser window is resized (currently the Transitions report is not re-drawn and will overflow or look small)

**Comment by tsteur** [comment]
* added the icon
* added help text & link
* couldn't reproduce an error when switching site.... added some code maybe it helps
* No we cannot resize it, it's a "miracle" it was possible to embed it at all

**Pull request merged**